dbms/src/Functions/LeastGreatest.h

```diff
@@ -120,5 +122,289 @@ class FunctionVectorizedLeastGreatest : public IFunction
+{
+public:
+    static constexpr auto name = "tidbLeastString";
+    explicit FunctionLeastString() = default;
+    ;
+    static FunctionPtr create(const Context & context [[maybe_unused]])
+    {
+        return std::make_shared<FunctionLeastString>();
+    }
+
+    String getName() const override { return name; }
+    bool isVariadic() const override { return true; }
+
+    void setCollator(const TiDB::TiDBCollatorPtr & collator_) override
+    {
+        collator = collator_;
+    }
+
+    size_t getNumberOfArguments() const override { return 0; }
+    bool useDefaultImplementationForNulls() const override { return true; }
+
+    DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override
+    {
+        if (arguments.size() < 2)
+            throw Exception(
+                fmt::format("Number of arguments for function {} doesn't match: passed {}, should be at least 2.", getName(), arguments.size()),
+                ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
+
+        for (const auto & argument : arguments)
+        {
+            if (!argument->isString())
+            {
+                throw Exception(
+                    fmt::format("argument type not string"),
+                    ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+            }
+        }
+        return std::make_shared<DataTypeString>();
+    }
+
+    void executeImpl(Block & block, const ColumnNumbers & arguments, size_t result) const override
+    {
+        size_t num_arguments = arguments.size();
+        if (num_arguments < 2)
+        {
+            throw Exception(
+                fmt::format("Number of arguments for function {} doesn't match: passed {}, should be at least 2.", getName(), arguments.size()),
+                ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
+        }
+        using impl = StringOperationImpl;
+        const auto * c0 = block.getByPosition(arguments[0]).column.get();
+        auto c_res = ColumnString::create();
+
+        for (size_t i = 1; i < num_arguments; ++i)
+        {
+            c_res = ColumnString::create();
+            if (i > 1)
+            {
+                c0 = block.getByPosition(result).column.get();
+            }
+            const auto * const c1 = block.getByPosition(arguments[i]).column.get();
+            const auto * c0_string = checkAndGetColumn<ColumnString>(c0);
+            const auto * c1_string = checkAndGetColumn<ColumnString>(c1);
+            const ColumnConst * c0_const = checkAndGetColumnConstStringOrFixedString(c0);
+            const ColumnConst * c1_const = checkAndGetColumnConstStringOrFixedString(c1);
+            if (c0_const && c1_const)
+            {
+                String res;
+                impl::constantConstant(c0_const->getValue<String>(), c1_const->getValue<String>(), collator, res);
+                block.getByPosition(result).column = block.getByPosition(result).type->createColumnConst(c0_const->size(), toField(res));
```

Comment:
I think we should scan all the columns beforing processing so that we can know if res is const column at the beginning.